Create multi-tenant DI sample (#4249)
Particular/docs.particular.net: samples/multi-tenant/di/sample.md

---

> title: Injecting tenant-aware components to the message handlers

Reviewer: ```suggestion
title: Injecting tenant-aware components to message handlers
```

> summary: Shows how to configure IoC to inject tenant-aware components to the message handlers

Reviewer: ```suggestion
summary: Shows how to configure IoC to inject tenant-aware components to message handlers
```

reviewed: 2019-02-07

> component: Core

Reviewer: A link to behaviors documentation as related would be a nice to have

---

## Introduction

> This sample shows how to configure the dependency injection infrastructure built into NServiceBus to instantiate and inject tenant-aware components. Such approach allows the code in the handler to focus on the actual business logic without having to deal with multi-tenant aspect of the problem.

Reviewer: ```suggestion
This sample shows how to configure the dependency injection infrastructure built-in into NServiceBus to instantiate and inject tenant-aware components. Such approach allows code in handlers to focus on the actual business logic without having to deal with multi-tenant aspects of the problem.
```

Author: I made it plural

This sample simulates a multi-tenant solution where the session provided to handlers is connected to individual tenant databases based on the value of a `tenant` header on the incoming message.

## Code Walk Through

### Creating the Behavior

The behavior will wrap handler invocations and:

 1. Open a session to the relevant tenant database.
 1. Make the session available to the message handlers.
 1. Commit or rollback the session depending on the outcome of the handler.
 1. Dispose the session.

snippet: unit-of-work-behavior

Note that because the user can't provide arguments when resolving the session instance, the session has to be initialized after construction. The `Initialize` method is defined on the concrete `MySession` class. It is not visible to the handler because the handler requires only the `ISession` interface.

### Registering the behavior

The following code is needed to register the behavior in the receive pipeline.

snippet: configuration

NOTE: The lifecycle has to be specified as `DependencyLifecycle.InstancePerUnitOfWork` in order to ensure there is a single instance of session created for each message being handled.

### Message handlers

> One of the benefits of a unit of work is that multiple handlers for the same message will share the same session and commit/rollback together. This is how the handlers look:

Reviewer: ```suggestion
One of the benefits of a unit of work is that multiple handlers for the same message will share the same session and commit/rollback together. This is how handlers look like:
```
